src/theory/quantifiers/sygus_inst.cpp

```diff
@@ -138,6 +138,100 @@ bool SygusInst::checkCompleteFor(Node q)
   return d_inactive_quant.find(q) != d_inactive_quant.end();
 }
 
+/*
+ * Find maximal ground terms in node n.
+ * Note: We do not collect Boolean ground terms since this would always cut off
+ * other non-Boolean terms.
+ *
+ * @param: stop_at_quant: Do not traverse down quantifiers.
+ */
+static void getMaxGroundTerms(
+    TNode n,
+    std::unordered_set<Node, NodeHashFunction>& terms,
+    std::unordered_set<TNode, TNodeHashFunction>& cache,
+    bool stop_at_quant = false)
+{
+  if (options::sygusInstGranularity() != options::SygusInstGranularity::MAXIMAL
+      && options::sygusInstGranularity() != options::SygusInstGranularity::BOTH)
+  {
+    return;
+  }
+
+  Node cur;
+  std::vector<TNode> visit;
+
+  visit.push_back(n);
+  do
+  {
+    cur = visit.back();
+    visit.pop_back();
+
+    if (cache.find(cur) != cache.end())
+    {
+      continue;
+    }
+    cache.insert(cur);
+
+    if (expr::hasBoundVar(cur) || cur.getType().isBoolean())
```

Comment:
Why is this check for Boolean type necessary?  Its possible that we could be doing QBF and have a large ground term of type Boolean.

Reply:
refactored as discussed off-line.